module/zfs/zfs_sa.c

```diff
@@ -337,7 +337,7 @@ zfs_sa_upgrade(sa_handle_t *hdl, dmu_tx_t *tx)
 	SA_ADD_BULK_ATTR(sa_attrs, count, SA_ZPL_SIZE(zsb), NULL,
 	    &zp->z_size, 8);
 	SA_ADD_BULK_ATTR(sa_attrs, count, SA_ZPL_GEN(zsb),
-	    NULL, &(ZTOI(zp)->i_generation), 8);
+	    NULL, &(ZTOI(zp)->i_generation), 4);
```

Comment:
After a close look, this is actually doing a write. You need to do a lookup using a uint64_t above the sa_bulk_lookup_locked(). And then use that value here. We cannot truncate the value in SA.